ls1intum/Artemis
# `Exam mode`: Load initial data of the exam live statistics

### Checklist
#### General


#### Server
#### Client

### Motivation and Context



To increase usability and performance, we also want to support initial loading of received actions. A user should not have to interact with the system. This should be done without any interaction.

### Description


The received actions are stored in the `ExamActionService`. 

Once the user reloads the page, we send a REST request to the server to get all the initial actions. These actions are than prepared and evaluated on the client. 

- Adjusted the init and update functionality of charts
- Added additional REST calls to receive the actions (`api/exam-monitoring/{examId}/load-actions`)

### Steps for Testing

Prerequisites:
- 1 Instructor
- Multiple Students
- 1 Exam with Exam Live Statistics enabled

1. Log in to Artemis
2. Create an Exam with Exam Live Statistics enabled
3. Register multiple Students
4. Participate with those students
5. View the Exam Live Statistics as instructor during the Exam
6. Reload the page (receiving and rendering the actions can take up to 15 seconds -> re-rendering + initial load)
7. Switch sections
8. Verify that the provided information are correct and the old actions are also loaded

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| average-actions-chart.component.ts | 76.92% | 96.29% |
| category-actions-chart.component.ts | 76.47% | 97.43% |
| total-actions-chart.component.ts | 80% | 97.29% |
| chart.component.ts | 70.37% | 88.23%* |
| exercise-chart.component.ts | 83.33% | 96.15% |
| exercise-group-chart.component.ts | 85.71% | 96.87% |
| exercise-navigation-chart.component.ts | 82.35% | 97.29% |
| exercise-submission-chart.component.ts | 82.35% | 97.29% |
| monitoring-chart.ts | 93.87% | 98.3% |
| exam-action.service.ts | 91.66% | 100% |
| exam-monitoring.component.ts | 86.66% | 100% |
| monitoring-overview.component.ts | 75% | 100% |
| ExamMonitoringScheduleService.java | 63% | 86%* |
| ExamActivityResource.java | n/a | 100% |

\* will be increased in https://github.com/ls1intum/Artemis/pull/5301
